### PR TITLE
nixos/onedrive: init

### PR DESCRIPTION
**Pull request by SRGOM**
### Documentation

https://github.com/NixOS/nixpkgs/pull/77734#issuecomment-575874225
 
  ###  Ideal Todo but I fail: 

* Use full path instead of just a dir name. 
Currently this module allows using only a directory name inside ~/.config. Should actually allow full directory paths. I believe that using paths as instance name is not allowed by systemd.

* Use `$XDG_CONFIG_HOME` instead of `.config`. The author of the main package- onedrive, uses .config and not `$XDG_CONFIG_HOME` in `https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md`. I would prefer using XDG but would go with .config because that's what the author seemingly uses, not sure what he's doing in code

** User abraunegg comments that XDG_CONFIG_HOME isn't a safe value to use anyway. 

### Todo that I wouldn't be able to spend time on... 
* Automated testing. I've never written a test and don't have the time or motivation unfortunately.

### Manual testing steps: 

1) Check out. 
2) Inside `imports` section, add a line with absolute path to the file `onedrive.nix`
3) Add `services.onedrive.enable = true; services.onedrive.package= <SOME RECENT CHANNEL>.onedrive`

### CODE REVIEW

**Comment by doronbehar** [comment]
Hey @SRGOM.

First of all, I'm not using onedrive, I just thought it's cool and that we should package it and I don't mind maintaining it much. But, upstream [provides a systemd service](https://github.com/abraunegg/onedrive/tree/master/contrib/systemd). Do we package it? If we do, I don't think we should add our own version of it. If we don't package it, let's start by adding it to our package and perhaps make sure everything here:

https://github.com/abraunegg/onedrive/tree/master/contrib

Will get packaged as well. BTW it needs a small version update ([see](https://github.com/abraunegg/onedrive/releases)) it could be nice to get on the way..

**Comment by SRGOM** [comment]
@doronbehar  thanks for responding. I have to confess :

1) I don't know much about systemd in general. 

2) I don't know how to package existing services in nixos (I will find out)

That said, this code allows multiple oned instances for the same user. So at least one of the two services is different. 

Also,  sorry I don't know what you mean by "adding it to our package". Don't modules live in diff hierarchy? (First time module writer ) 

I may not be able to spend effort to package other things, sorry. I believe a service is critical hence adding.

**Comment by SRGOM** [comment]
@doronbehar I did a `find` for `.service` files in the nixpkgs repository, and there's no such file. So I guess you meant just copy paste the service from upstream into a `nix` style service. Got it.

**Comment by SRGOM** [comment]
@doronbehar I have addressed your comment and incorporated the service from onedrive upstream. Any other comments you'd like to add?

**Comment by SRGOM** [comment]
@jonringer Since you too reviewed (and approved) https://github.com/NixOS/nixpkgs/pull/73360, may I further abuse your hospitality by requesting you to review this too? I still have to add documentation but the service part is done (as best as I can- please see the Ideal Todo part in my original post)

**Comment by aanderse** [comment]
@SRGOM I'm pretty sure @doronbehar meant you should use `systemd.packages` instead of writing (or copy+pasting) `systemd` unit files. Do a quick grep through nixpkgs and you'll see when upstream has a `systemd` unit file we prefer that.

Also please keep to a consistent 2 spaces per level of indentation.

**Comment by doronbehar** [comment]
> I don't know how to package existing services in nixos (I will find out)

When I said "package files from upstream", I meant that we need to make sure our `onedrive` **package definition** should include the files from https://github.com/abraunegg/onedrive/tree/master/contrib . I assumed you are somewhat familiar with the nixpkgs packages' build system (this is where most people start) - sometimes, packages build systems "install" such files to the the prefix which is `$out` in our case. @SRGOM If almost every word / phrase I'm using here isn't understood to you, please read about this in the manual, as I can't fill for you all of the necessary gaps here and now.

In general, my comment comes from the question: What does this service / module does that the package doesn't provide by it self? And if the package should provide something, I think it'd be better to leave it that way.

**Comment by SRGOM** [comment]
Alright, I'm going to address your comments as soon as I can (I'll probably need to read up quite a bit). But quickly, @doronbehar- my code allows running the service for an unlimited number of onedrive  accounts for the same user with very little effort. Just list all the config dirs in ~/.config/onedrive-launcer and onedrive services will be instantiated for all those accounts.

**Comment by SRGOM** [comment]
Actually I just re-read both your comments and yeah, I happen to (fortunately) understand nix packaging. I understand (now)  how services in Nix are handled, just like bins are moved to $out/bin, services are moved to $out/....servies path.... Fine. 

With that understanding I want to say that the module I am submitting is very different and cannot make use of anything but adapted code from upstream. 

In my code- I have two things- one is an instantiable "onedrive@" service, which instantiates on config directories. The other is a launcher service which launches each of these instantiated services.

The upstream code cannot be reused for that purpose without adapting.

@aanderse I will fix spacing.

File: nixos/modules/programs/onedrive.nix [hunk 1]
```diff
@@ -0,0 +1,73 @@
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.programs.onedrive;
+
+	onedriveLauncher =  pkgs.writeShellScriptBin
+		"onedrive-launcher"
+		''
+			# XDG_CONFIG_HOME is not recognized in the environment here.
+			if [ -f $HOME/.config/onedrive-launcher ]
```
Comment (ianmjones) [inline nixos/modules/programs/onedrive.nix]: Is this a commonly used file name used for this?

If not, maybe this should be an option that uses this name as the default?

Better still, how about using an option that allows you to list config directories? Maybe `services.onedrive.configDirs` with the default value being `[ "~/.config/onedrive" ]`?

File: nixos/modules/programs/onedrive.nix [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+			then
+				# Hopefully using underscore boundary helps locate variables
+				for _onedrive_config_dirname_ in $(cat $HOME/.config/onedrive-launcher | grep -v '[ \t]*#' )
+				do
+					systemctl --user start onedrive@$_onedrive_config_dirname_
```
[the remark this answers is not in the document]
Comment (SRGOM) [reply nixos/modules/programs/onedrive.nix]: I will document that- First run and setting up a config are manual steps, taken per user.

File: nixos/modules/programs/onedrive.nix (hunk 1 above)
Comment (SRGOM) [reply nixos/modules/programs/onedrive.nix]: I am not too keen on doing things at the "system" level, this is thing that should be done at user level. 

Basically the user does what he always does- initialize with oauth and config, and then the service runs as is. Except this time, you can have multiple accounts.

The filename- "onedrive-launcher" is not standard. This is our creation. Once again, it's per user and hence probably good to keep standard and also not customizable per system (if it were customizable per user that would have been fine)

Comment (ianmjones) [reply nixos/modules/programs/onedrive.nix]: Gotcha, makes sense.

File: nixos/modules/programs/onedrive.nix [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+	systemdTargets = builtins.listToAttrs
+	(
+		map
+			( st:{ name = st; value = st + ".target" ; } )
+			systemdTargetList
+	);
+
+
+in {
+  options.programs.onedrive = {
```
[the remark this answers is not in the document]
Comment (SRGOM) [reply nixos/modules/programs/onedrive.nix]: I addressed your comment.

**Comment by SRGOM** [comment]
Documentation screenshot-
DeepinScreenshot_select-area_20200117204941

**Comment by SRGOM** [comment]
I am also adding `config.services.onedrive.package` to `environment.systemPackages`- this is because first run needs to be manual for setting up oauth by visiting link on the browser.

**Comment by ianmjones** [comment]
@SRGOM Looks great!

**Comment by aanderse** [comment]
Just out of curiosity... is this module something which might be better suited for `home-manager` instead?

**Comment by SRGOM** [comment]
I'm not sure what defines the boundaries and shared regions but in this particular case, if multiple users use a single OneDrive account, no configuration on their part besides the initial oauth is required once the sysadmin enables the service. That's a strong argument in favor of having a "global " service.

That said, I'm always curious why somebody would use home manager at all? Having packages on a multi user system, sure- but it's a bad idea as far as config management is concerned - you cannot edit a file and let things apply on the fly (unlike the dotfile + stow approach). And you don't have a "stateless" config directory anyway- things are very stateful.

**Comment by SRGOM** [comment]
DeepinScreenshot_select-area_20200118124150

**Comment by abraunegg** [comment]
> * Use `$XDG_CONFIG_HOME` instead of `.config`. The author of the main package- onedrive, uses .config and not `$XDG_CONFIG_HOME` in `https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md`. I would prefer using XDG but would go with .config because that's what the author seemingly uses, not sure what he's doing in code

@SRGOM 

Great question. Based on extensive testing and fixing issues / bugs from the original client, many distro's do not set correctly the `XDG_CONFIG_HOME` value, combined with that if the system is entirely headless (no X11) this value will also not exist.

In the code itself (config.d) there is the following:
```
// Determine the base directory relative to which user specific configuration files should be stored.
if (environment.get("XDG_CONFIG_HOME") != ""){
	log.vdebug("configDirBase: XDG_CONFIG_HOME environment variable set");
	configDirBase = environment.get("XDG_CONFIG_HOME");
} else {
	// XDG_CONFIG_HOME does not exist on systems where X11 is not present - ie - headless systems / servers
	log.vdebug("configDirBase: WARNING - no XDG_CONFIG_HOME environment variable set");
	configDirBase = homePath ~ "/.config";
}
```

This allows to check if `XDG_CONFIG_HOME` is set, but if not use `~/.config` 

If NixOS is reliably setting `XDG_CONFIG_HOME` 100% of the time, and this is the norm for the distro from a documentation perspective, I would rather you follow your documentation standards than 'break' that just for this application.

Other than that, the documentation changes / additions look fine to me.

**Comment by SRGOM** [comment]
@abraunegg Thank you, clearly you've spent time worrying about stability. That `if else` check makes me want to use XDG_CONFIG_HOME but (and I forgot to update my original post) XDG_CONFIG_HOME isn't available in the systemd service. 

> Entirely headless => no XDG_CONFIG_HOME
Interesting.

**Review by infinisil — CHANGES_REQUESTED** [review]
Not sure about this module, having the user write a config file to their home directory to configure something is not really in the spirit of NixOS at all (which is about declarative system config, not stateful user services). Perhaps [home-manager](https://github.com/rycee/home-manager) is a better place for such a module, since it's focused on single users, allowing you to control home directory paths too.

**Comment by SRGOM** [comment]
You mention "not sure". Is that a final in which case I will close because I have no motivation to write a home manager module. As I noted above, and I'm not being snarky, 

> That said, I'm always curious why somebody would use home manager at all? Having packages on a multi user system, sure- but it's a bad idea as far as config management is concerned - you cannot edit a file and let things apply on the fly (unlike the dotfile + stow approach). And you don't have a "stateless" config directory anyway- things are very stateful.

Also when I open my ~/.ssh/config in vim, `ProxyCommand` is colored but `PoxyCommand` isn't. Not something home manager can give yopu.

This isn't the occasion to be snarky but if unofficial solution here seems a bit off. 

Thank you for spending the time to look and for maintaining NixOS. (I hope I didn't sound ungrateful with my above comments, it's just that I feel a bit strongly about home manager I feel that HM is the nail to the (amazing) Nix hammer.

**Comment by ianmjones** [comment]
What if there was a compromise that meant the system admin could declare which users OneDrive is enabled for?

```
services.onedrive.package = pkgs.onedrive;

services.onedrive.users."ian" = {
  enable = true;
  configDirs = [ ".config/onedrive" ".config/office365" ];
};
```

This means the `systemd` services are controlled by the system administrator rather than users.

The module would only set up the services where declared users have a config file in one of the prescribed config directories.

`services.onedrive.users.<name?>.configDirs` would be optional, defaults to value of `services.onedrive.configDirs`.

`services.onedrive.configDirs` would be optional, defaults to `[ ".config/onedrive" ]`.

`configDirs` entries are always evaluated relative to the user's home dir.

**Comment by Infinisil** [comment]
A general problem with NixOS modules that control user-level things is that users can't control it unless they're a super-user: If you want to enable such a service, you need to be able to call `nixos-rebuild switch`.

In contrast, home-manager allows you to do that, because it runs on a per-user basis, with its config file in `~/.config/nixpkgs/home.nix`. And for the people who have single-user systems, they can also [run home-manager as part of NixOS](https://rycee.gitlab.io/home-manager/index.html#sec-install-nixos-module), allowing you to configure home-manager from your `configuration.nix`:

```nix
{
  home-manager.users.alice = {
    services.onedrive.enable = true;
  };
}
```

home-manager is nowadays pretty much the standard way to configure user-level things. Yes it's a third-party tool, but I think we should embrace this, because nixpkgs is really quite big already. Once NixOS/rfcs#42 is finished it will also allow easier usage of third-party Nix code and libraries.

**Comment by abraunegg** [comment]
Given I have absolutely no knowledge of NixOS, please can you keep this in mind when reading below & making your decision as how best to run this under your OS / distribution:

1. If you run 'onedrive' as a 'super user' (eg root) either via cron or systemd, then all the files downloaded from OneDrive will have a UID/GID = this user. The issue here is that 'many' folk run this as a system service (see https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#onedrive-service-running-as-a-non-root-user-via-systemd-without-notifications-or-gui). Example - you run as systemd, but configure the `sync_dir` to be `/path/to/my/files` - essentially you will get access / permission issues and problems when attempting to read / write the files.
2. Whilst the binary is installed with root privileges, standard users can run & configure the application as to how they want to 'use it'. Essentially the application 'has' to run as the user that intends to access / use the files. This allows then 'many' users on the same system, at the same time can use the same binary, but with their unique configuration to access OneDrive & sync in the manner that they want.
3. The `~/.config/onedrive` user path serves the following purposes:
    * Storing the 'configuration' file if you want to 'modify' the application defaults (see: https://github.com/abraunegg/onedrive/blob/master/docs/USAGE.md#configuration)
    * Storing the 'application sync cache' data
    * Storing a 'hash' to detect configuration changes
    * Storing the users O2Auth information for accessing the OneDrive service
4.  The config path is configurable via the CLI (or by modifying the systemd / initd scripts) by using `--confdir` - if `~/.config/onedrive`  if not applicable, then store elsewhere, but ensure that the `--confdir` is used at each application runtime.


@ianmjones
> This means the `systemd` services are controlled by the system administrator rather than users.

See above - unless you enable the right user service, then it will cause issues when downloading files as a non super / system admin user.

**Comment by ianmjones** [comment]
@abraunegg said...

> unless you enable the right user service, then it will cause issues when downloading files as a non super / system admin user.

Yup, this module for NixOS only creates user level systemd units, and only starts the service for the user if they have a recognised config dir for onedrive.

**Comment by SRGOM** [comment]
Thank you, I appreciate the good faith discussion. I can see how home manager shines here. I want to add one thing though (I hope I'm not answering a non-question)-

Unlike other distros, NixOS allows users to install their own packages. So I can see how it might be a concern that admins are required for a service. But that's true for a lot of other services, (redshift comes to mind). So this service isn't much different from others really. 

Also, just to give an idea of how this interacts with config files. 


Assume disabled service. Nothing happens.

Assume enabled service. If I as a user have .config/OneDrive, then I likely want the service to run anyway. But the module gives you an option to not run the individual service if you do not wish.

Basically once the admin enables the service, only the oneshot onederivelauncher service starts mandatorily . Everything else is trivially manageable by the user.

**Comment by SRGOM** [comment]
@ianmjones Sorry I missed your previous comment- I believe the concern  infinisil's concern here is the lack of flexibility (please correct me if not). As such my previ0us comment tries to address that.

Was your concern in response to that. If not, does my previous comment seem to address what your motivation might have been? Happy to answer more

**Comment by ianmjones** [comment]
@SRGOM No worries, my [previous comment](https://github.com/NixOS/nixpkgs/pull/77734#issuecomment-576613365) was indeed trying to address @Infinisil's original comment about not liking the requirement of users adding config files to their home dir that are specific to this module (that's how I read their comment). As you know, I too raised concerns about the `~/.config/onedrive-launcher` file, it doesn't feel right.

It also doesn't feel right to me that we're currently needing to install a third party channel (home-manager) in order to get per user declarative config, hence my suggested compromise.

However, I do like the format that @Infinisil's follow-up home-manager example used for enabling the onedrive service on a per user basis, so I'm going to give home-manager a try.

It is a shame though that we can't have a OneDrive service module that enables users to optionally use it simply by having the standard config in place, and doing the one-time manual authentication to confirm access.

**Comment by SRGOM** [comment]
I actually think that allowing the user the flexibility of turning on and off sync without a system rebuild is a positive. I believed that your concern arose from it not being a package defined standard but something my module created ( I share that but I don't think it's a deal breaker).

File: nixos/modules/services/networking/onedrive.nix [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+    ''
+  ;
+
+  # Poor man's type/typo safety
+  systemdTargetList = [
+    "network-online"  
+    "multi-user"
+  ];
+
+  systemdTargets = builtins.listToAttrs
```
Comment (peterhoeg) [inline nixos/modules/services/networking/onedrive.nix:28]: this bit is needlessly complicated but it actually doesn't matter for reasons outlined below as it should just go.

File: nixos/modules/services/networking/onedrive.nix [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+  ### Interface 
+
+  options.services.onedrive = {
+    enable = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = "Enable OneDrive service";
+    };
+
+     package = lib.mkOption {
```
Comment (peterhoeg) [inline nixos/modules/services/networking/onedrive.nix:49]: Do we have multiple onedrive packages to choose from or does the single package have a lot of compile time flags that people would want to set? If not, get rid of this.

File: nixos/modules/services/networking/onedrive.nix [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+  config = lib.mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+
+    systemd.user.services."onedrive@" = {
+      description = "Onedrive sync service";
+
+      serviceConfig = {
+        Type = "simple";
+        ExecStart = ''
+          ${pkgs.bash}/bin/bash -c "${cfg.package}/bin/onedrive --monitor --verbose --confdir=$HOME/.config/%i"
```
Comment (peterhoeg) [inline nixos/modules/services/networking/onedrive.nix:70]: 1. Why are you wrapping this in `bash`?
2. The idiomatic systemd way of referencing the home directory is `%h`, not `$HOME` inside a unit.

File: nixos/modules/services/networking/onedrive.nix [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+          '';
+        Restart="on-failure";
+        RestartSec=3;
+        RestartPreventExitStatus=3;
+      };
+    };
+
+    systemd.user.services.onedriveLauncher = {
+      wantedBy = [ systemdTargets."multi-user" ];
```
Comment (peterhoeg) [inline nixos/modules/services/networking/onedrive.nix:79]: This doesn't do what you think - a systemd user session doesn't know anything about `multi-user.target`. That is a system level target, so in effect this is never pulled in.

File: nixos/modules/services/networking/onedrive.nix [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+      after = [ systemdTargets."network-online" ];
```
Comment (peterhoeg) [inline nixos/modules/services/networking/onedrive.nix:80]: Same thing here.

File: nixos/modules/services/networking/onedrive.nix [hunk 9]
```diff
@@ -0,0 +1,88 @@
+{ config, lib, pkgs, ... }:
+let
+  cfg = config.services.onedrive;
+
+  onedriveLauncher =  pkgs.writeShellScriptBin
```
Comment (peterhoeg) [inline nixos/modules/services/networking/onedrive.nix:5]: I personally don't like the use of a launcher like this as you cannot individually start and stop the onedrive instances as the launcher will process all you have configured. IMHO, it's far better to just create the `onedrive@` unit and have the user activate what needs to be activated. If multiple instances are expected, you could wrap those in a `onedrive.target` and have `default.target` pull in `onedrive.target`. That gives you a way to start them all at one go as well as start/stop them individually.

File: nixos/modules/services/networking/onedrive.nix [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+### Implementation
+
+  config = lib.mkIf cfg.enable {
+    environment.systemPackages = [ cfg.package ];
+
+    systemd.user.services."onedrive@" = {
+      description = "Onedrive sync service";
+
+      serviceConfig = {
+        Type = "simple";
```
Comment (peterhoeg) [inline nixos/modules/services/networking/onedrive.nix:68]: See if it works with `exec` instead.

**Review by peterhoeg — CHANGES_REQUESTED** [review]

**Comment by devhell** [comment]
Is there a status on this PR? Upstream OneDrive for Linux client author seems to believe it's been rejected?

**Comment by peterhoeg** [comment]
Definitely not rejected but there are a few things that need to be addressed.

**Comment by SRGOM** [comment]
I am not an expert in linux systemd related aspects. That combined with my computer's internet issues which require me to manually start internet after every login make it hard for me to understand if the service has started at login. My original code did work fine for me (I had good internet then)

That said, I want to thank @peterhoeg for the review and also clarify one bit about how the launcher works- 

You can definitely start and stop a single service (I do it all the time to save bandwidth). What the launcher adds is the ability to start all your onedrive instances in one go. So the oneDriveLauncher is a good thing. 

I am not sure when I will get time again to complete this PR. If someone else wants to go ahead with this, please feel free .

**Comment by abraunegg** [comment]
@devhell

> Is there a status on this PR? Upstream OneDrive for Linux client author seems to believe it's been 
rejected?

Updated the documentation (https://github.com/abraunegg/onedrive/commit/ed1d13bc1a06bbaa5baa4ddb072794b126018c4f) to reflect that this is not rejected, but needs work

**Comment by SRGOM** [comment]
@peterhoeg The way you made suggestions makes it sound like you understand systemd well, could you kindaly suggest a way to quickly test start-up of user services- I wonder if `kill -9 -1`, and `kill -9 0` suffice?

File: nixos/modules/services/networking/onedrive.nix (hunk 10 above)
Comment (SRGOM) [reply nixos/modules/services/networking/onedrive.nix:68]: Could you explain what you mean by this? I did try exec but looking further there is no `Type=exec`. All types- https://www.freedesktop.org/software/systemd/man/systemd.service.html

**Comment by peterhoeg** [comment]
I'm sorry, I don't understand what you are asking. What exactly are you trying to do?

File: nixos/modules/services/networking/onedrive.nix (hunk 10 above)
Comment (SRGOM) [reply nixos/modules/services/networking/onedrive.nix:68]: I've left the type as is. Couldn't find an `exec` type.

File: nixos/modules/services/networking/onedrive.nix (hunk 9 above)
Comment (SRGOM) [reply nixos/modules/services/networking/onedrive.nix:5]: This launcher only helps in one way- starting all (say 4) onedrive service instances. This launcher is auto-started because it is `wantedBy` `default.target`. One can manually start the launcher. 

One can also manually start and stop individual onedrive client instance services.

File: nixos/modules/services/networking/onedrive.nix (hunk 8 above)
Comment (SRGOM) [reply nixos/modules/services/networking/onedrive.nix:80]: Replaced `multi-user.target` with `default.target`, needed to start user service upon login

File: nixos/modules/services/networking/onedrive.nix (hunk 7 above)
Comment (SRGOM) [reply nixos/modules/services/networking/onedrive.nix]: Replaced with `default.target`, which is required to aut0-start user services upon login.

File: nixos/modules/services/networking/onedrive.nix (hunk 6 above)
Comment (SRGOM) [reply nixos/modules/services/networking/onedrive.nix]: Done like you asked. No wrapping in bash, use `%h` now

File: nixos/modules/services/networking/onedrive.nix (hunk 4 above)
Comment (SRGOM) [reply nixos/modules/services/networking/onedrive.nix:28]: Left this in because your objection probably wasn't to this code in itself.

**Comment by SRGOM** [comment]
@Peterhoeg hi, let me know if you need further changes here. I've spent some time today and I can spend more time fixing this since I'm (again) familiar with the code. The service starts perfectly for me upon reboot and the subsequent login.

File: nixos/modules/services/networking/onedrive.nix (hunk 5 above)
Comment (SRGOM) [reply nixos/modules/services/networking/onedrive.nix:49]: I've left this in because onedrive upstream moves very fast, and there might be times where one is subscribed to a release channel but would like to use the latest package. For example in fact reportedly the version in stable has problems with onedrive business, and this pr-  https://github.com/NixOS/nixpkgs/pull/89198 upgrades the version. 

(I know it's possible to use a specific service from a specific channel/commit but that's a tiny bit more work than using a package from another channel)

File: nixos/modules/services/networking/onedrive.nix [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+        ExecStart = ''
+          ${cfg.package}/bin/onedrive --monitor --verbose --confdir=%h/.config/%i
+        '';
+        Restart="on-failure";
+        RestartSec=3;
+        RestartPreventExitStatus=3;
+      };
+    };
+
+    systemd.user.services.onedriveLauncher = {
```
Comment (peterhoeg) [inline nixos/modules/services/networking/onedrive.nix:64]: Minor nitpick - following our naming convention it would be `onedrive-launcher`.

File: nixos/modules/services/networking/onedrive.nix [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+      wantedBy = [ "default.target" ];
+      after = [ "network-online.target" ];
```
Comment (peterhoeg) [inline nixos/modules/services/networking/onedrive.nix:66]: `network-online.target` is not available to user services.

**Review by peterhoeg — CHANGES_REQUESTED** [review]

**Comment by asdf8dfafjk** [comment]
@peterhoeg I have continued this work in https://github.com/NixOS/nixpkgs/pull/91765 due to GitHub login issues.
